softlayer/vm/ubuntu.go

```diff
@@ -0,0 +1,324 @@
+		return nil, err
+	}
+
+	privateInterface := Interface{
+		Name:           fmt.Sprintf("%s%d", privateComponent.Name, privateComponent.Port),
+		Auto:           true,
+		AllowHotplug:   true,
+		Address:        privateComponent.PrimaryIPAddress,
+		Netmask:        subnet.Netmask,
+		Gateway:        subnet.Gateway,
+		DefaultGateway: (publicComponent.PrimaryIPAddress == "" && nw.IsDefaultGateway()),
+		Routes:         SoftlayerPrivateRoutes(subnet.Gateway),
+	}
+	interfaces := []Interface{privateInterface}
+
+	if publicComponent.PrimaryIPAddress != "" {
+		for _, s := range publicComponent.NetworkVLAN.Subnets {
+			if s.Contains(publicComponent.PrimaryIPAddress) {
+				subnet = s
+				break
+			}
+		}
+		publicInterface := Interface{
+			Name:           fmt.Sprintf("%s%d", publicComponent.Name, publicComponent.Port),
+			Auto:           true,
+			AllowHotplug:   true,
+			Address:        publicComponent.PrimaryIPAddress,
+			Netmask:        subnet.Netmask,
+			Gateway:        subnet.Gateway,
+			DefaultGateway: nw.IsDefaultGateway(),
+		}
+		interfaces = append(interfaces, publicInterface)
+	}
+
+	return interfaces, nil
+}
+
+func (u *Ubuntu) manualInterfaces(networkComponents VirtualGuestNetworkComponents, networks Networks) ([]Interface, error) {
+	privateComponent := networkComponents.PrimaryBackendNetworkComponent
+	publicComponent := networkComponents.PrimaryNetworkComponent
+
+	interfaces := []Interface{}
+	for networkName, nw := range networks {
+		if subnet, err := privateComponent.NetworkVLAN.Subnets.Containing(nw.IP); err == nil {
+			intf := Interface{
+				Name:           fmt.Sprintf("%s%d:%s", privateComponent.Name, privateComponent.Port, networkName),
+				Auto:           true,
+				AllowHotplug:   true,
+				Address:        nw.IP,
+				Netmask:        subnet.Netmask,
+				Gateway:        subnet.Gateway,
+				DefaultGateway: nw.IsDefaultGateway(),
+				Routes:         SoftlayerPrivateRoutes(subnet.Gateway),
+			}
+
+			interfaces = append(interfaces, intf)
+			continue
+		}
+
+		if subnet, err := publicComponent.NetworkVLAN.Subnets.Containing(nw.IP); err == nil {
+			intf := Interface{
+				Name:           fmt.Sprintf("%s%d:%s", publicComponent.Name, publicComponent.Port, networkName),
+				Auto:           true,
+				AllowHotplug:   true,
+				Address:        nw.IP,
+				Netmask:        subnet.Netmask,
+				Gateway:        subnet.Gateway,
+				DefaultGateway: nw.IsDefaultGateway(),
+			}
+
+			interfaces = append(interfaces, intf)
+			continue
+		}
+
+		return nil, errors.New("manual subnet not found")
+	}
+
+	return interfaces, nil
+}
+
+func (i Interfaces) Configuration() ([]byte, error) {
+	buf := &bytes.Buffer{}
+
+	t := template.Must(template.New("network-interfaces").Parse(ETC_NETWORK_INTERFACES_TEMPLATE))
+	err := t.Execute(buf, i)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+const ETC_NETWORK_INTERFACES_TEMPLATE = `# Generated by softlayer-cpi
```

Comment:
This is missing changes related to source based policy routing for public networks. It should be in my branch.